rbd: add support for changed block tracking RPCs

Rakshith-R opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #
rbd: add support for changed block tracking RPCs

This commit adds support for GetMetadataAllocated
and GetMetadataDelta RPCs.
[CSI Snapshot Metadata Service RPCs](https://github.com/container-storage-interface/spec/blob/master/spec.md#snapshot-metadata-service-rpcs)

Updates: #5346 

**Checklist:**

* [ ] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [ ] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

Rakshith-R committed:
rbd: add support for changed block tracking RPCs

This commit adds support for GetMetadataAllocated
and GetMetadataDelta RPCs.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
doc: add release note for changed block tracking feature

Signed-off-by: Rakshith R <rar@redhat.com>

Nikhil-Ladha commented on internal/rbd/controllerserver.go:
```diff
@@ -1709,3 +1709,67 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
 	return &csi.ControllerUnpublishVolumeResponse{}, nil
 }
+
+// GetSnapshotMetadata streams the allocated block metadata for a snapshot.
```
```suggestion
// GetMetadataAllocated streams the allocated block metadata for a snapshot.
```

Nikhil-Ladha commented on internal/rbd/controllerserver.go:
```diff
@@ -1709,3 +1709,67 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+	req *csi.GetMetadataDeltaRequest,
+	stream csi.SnapshotMetadata_GetMetadataDeltaServer,
+) error {
+	ctx := context.WithValue(stream.Context(), "Base-Snapshot-ID", req.GetBaseSnapshotId())
```
`Base Snapshot ID`

Nikhil-Ladha commented on internal/rbd/controllerserver.go:
```diff
@@ -1709,3 +1709,67 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+
+	targetSnapshotID := req.GetTargetSnapshotId()
+	if targetSnapshotID == "" {
+		return status.Error(codes.InvalidArgument, "base snapshot ID cannot be empty")
```
```suggestion
		return status.Error(codes.InvalidArgument, "target snapshot ID cannot be empty")
```

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1709,3 +1709,67 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
 
 	return &csi.ControllerUnpublishVolumeResponse{}, nil
 }
+
```
Lets move this to a new file as its a new functionality, This fine is already having 1770 lines of code.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1709,3 +1709,67 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+
+	rbdSnap, err := genSnapFromSnapID(ctx, snapshotID, cr, req.GetSecrets())
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
```
what about the NotFound error?

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,263 @@
+	maxResults int32) error {
+	image, err := rbdSnap.open()
+	if err != nil {
+		return err
```
return GRPC error message

Madhu-1 replied:
Take care of this in other places in this function

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,263 @@
+
+	changedBlocks := make([]*csi.BlockMetadata, 0)
+	diffIterateByIDConfig := librbd.DiffIterateByIDConfig{
+		Offset: uint64(startingOffset),
```
No validation is required for startingOffSet, what is the error message returned for invalid offSet?

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,263 @@
+			case <-stream.Context().Done():
+				return streamDoneErrCode
+			default:
+				changedBlocks = append(
```
Dont we get into the memory problem, If we have larger changedBlocks as we keep on appending to it?

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,263 @@
+				return sendResponseErrCode
+			}
+			// Reset the changedBlocks slice for the next batch
+			changedBlocks = make([]*csi.BlockMetadata, 0)
```
Can we have a fixed number of memory allocated per batch so that we dont get into OOM problem?

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,263 @@
+	"fmt"
+
+	"github.com/ceph/ceph-csi/internal/util/log"
+	librbd "github.com/ceph/go-ceph/rbd"
```
a new line is missing

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,263 @@
+			return nil
+		case sendResponseErrCode:
+			// An error occurred while sending the response.
+			log.ErrorLog(ctx, "failed to send response: %w", err)
```
log as %s, take care of it in other places as well

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,263 @@
+			// The error does not implement ErrorCode, return a generic internal error.
+			log.ErrorLog(ctx, "failed to get diff: %w", err)
+
+			return status.Error(codes.Internal, fmt.Sprintf("failed to get diff: $w", err))
```
cant we use status.Errorf here?

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,263 @@
+			// An unexpected error occurred.
+			log.ErrorLog(ctx, "failed to get diff: %w, error code: %d", err, errno)
+
+			return status.Error(codes.Internal, fmt.Sprintf("failed to get diff: %w, error code: %d", err, errno))
```
use status.Errorf

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,263 @@
+		}
+	}
+	// If no error, we successfully completed the diff operation.
+	if len(changedBlocks) > 0 {
```
Can we have a function to send the response and can be reused here and above like

```suggestion
send := func() error {
		if len(changedBlocks) == 0 {
			return nil
		}
		log.DebugLog(ctx, "Sending %d changed blocks to the stream", len(changedBlocks))
		if err := stream.Send(&csi.GetMetadataAllocatedResponse{
			BlockMetadataType:   csi.BlockMetadataType_VARIABLE_LENGTH,
			VolumeCapacityBytes: rbdSnap.VolSize,
			BlockMetadata:       changedBlocks,
		}); err != nil {
			log.ErrorLog(ctx, "failed to send response: %v", err)
			return status.Error(codes.Internal, fmt.Sprintf("failed to send response: %v", err))
		}
		changedBlocks = changedBlocks[:0] // Reset slice
		return nil
	}
```

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1709,3 +1709,67 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+
+	baseRBDSnap, err := genSnapFromSnapID(ctx, baseSnapshotID, cr, req.GetSecrets())
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
```
Need to handle NotFound?

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1709,3 +1709,67 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+
+	targetRBDSnap, err := genSnapFromSnapID(ctx, targetSnapshotID, cr, req.GetSecrets())
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
```
Need to handle NotFound?

Madhu-1 commented on internal/rbd/identityserver.go:
```diff
@@ -73,6 +73,19 @@ func (is *IdentityServer) GetPluginCapabilities(
 
+	// RBDSnapDiffByID is used within the SnapshotMetadata service
+	rbdSnapDiffByIDSupported, err := features.SupportsRBDSnapDiffByID()
+	if err == nil && rbdSnapDiffByIDSupported {
```
This is same as `SupportsGroupSnapGetInfo` (error is not logged) can we log the error for debug purpose?

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,263 @@
+	return nil
+}
+
+func (rbdSnap *rbdSnapshot) getMetadataDelta(
```
This looks similar to above and have lot of duplication . Can we rewrite this to avoid duplication of the core logic? That will help us to keep the outer implementation simple

Rakshith-R commented:
Thanks for the quick detailed review

The new main modifications are:
- moved the new calls to snap_diff.go file
- debug logs for feature support check errors
- refactored to have a single flow with sendStreamResponse handling which response to send allocated or delta
- modified return statement and ensured they are formatted and have correct error codes
- added validation for req parameters
- added check for snapshot's existence + returns correct error code now
- added additional checks for startingOffset (add out of range error code as specified in the [spec](https://github.com/container-storage-interface/spec/blob/master/spec.md#getmetadataallocated-errors))
- additional checks for maxResults and internal value to limit memory usage ([spec](spec allows us to send lower number of blocks as well))

failing golang lintci is being handled in https://github.com/ceph/ceph-csi/pull/5351

Please take a look again

iPraveenParihar commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,344 @@
+	defer cr.DeleteCredentials()
+
+	snapshotID := req.GetSnapshotId()
+	if snapshotID == "" {
```
not needed? already handled in `validateMetadataAllocatedReq`

nixpanic commented on internal/rbd/features/features.go:
```diff
@@ -29,8 +29,11 @@ import (
 	errGroupGetSnapInfo       error
+	errRBDSnapDiffByID        error
 	groupGetSnapInfoSupported = false
+	rbdSnapDiffByIDSupported  = false
```
nit: I prefer to keep the related variables grouped together, like:
```go
	groupGetSnapInfoOnce      sync.Once
	errGroupGetSnapInfo       error
	groupGetSnapInfoSupported = false

	rbdSnapDiffByIDOnce       sync.Once
	errRBDSnapDiffByID        error
	rbdSnapDiffByIDSupported  = false
```

nixpanic commented on internal/rbd/features/features.go:
```diff
@@ -56,3 +59,25 @@ func SupportsGroupSnapGetInfo() (bool, error) {
 	return groupGetSnapInfoSupported, errGroupGetSnapInfo
 }
+
+func SupportsRBDSnapDiffByID() (bool, error) {
```
every public/exported function should have a comment above it describing what it does.

nixpanic commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,344 @@
+
+	// streamDoneErrCode is a custom error code used to indicate that the
+	// stream was closed by the client.
+	streamDoneErrCode = -1
```
Is there a reason to not use `google.golang.org/grpc/codes` constants? Or, maybe base it on return values of `read()`, like `EOF` and so on?

nixpanic commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,344 @@
+
+	err = image.DiffIterateByID(diffIterateByIDConfig)
+	if err != nil {
+		errno, ok := err.(interface{ ErrorCode() int })
```
Isn't there a go-ceph error type that can be used for typecasting instead of an anonymous interface? Or, at least move the type-casting out into it's own function with a descriptive comment.

nixpanic commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,344 @@
+}
+
+// GetMetadataAllocated streams the allocated block metadata for a snapshot.
+func (cs *ControllerServer) GetMetadataAllocated(
```
Please don't mix functions for rbdImage and ControllerServer in the same file. Move the ControllerServer SMS functions (and helpers) into a separate file.

Eventually, after much more refactoring, the ControllerServer should only use the (internal) rbdManager API and opaque Volume type.

nixpanic commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,344 @@
+		return status.Error(codes.InvalidArgument, "snapshot ID cannot be empty")
+	}
+
+	rbdSnap, err := genSnapFromSnapID(ctx, snapshotID, cr, req.GetSecrets())
```
Use the `rbdManager` to get this, it should be(come) the only entry point to get Volumes and Snapshots. You will need to extend the Snapshot interface/type for the diff/metadata in a clean way too.

nixpanic commented on internal/rbd/snapshot.go:
```diff
@@ -400,3 +400,31 @@ func (rbdSnap *rbdSnapshot) SetVolumeGroup(ctx context.Context, cr *util.Credent
+	}
+	defer vol.Destroy(ctx)
+
+	image, err := vol.open()
```
Is there really way to get the `ParentInfo` through `rbdSnap`? Re-using `conn.Creds` is dangerous, as it may get free'd on the `vol.Destroy()`.

Rakshith-R replied:
> Is there really way to get the `ParentInfo` through `rbdSnap`? Re-using `conn.Creds` is dangerous, as it may get free'd on the `vol.Destroy()`.

The image underneath rbdSnap object is actually the parent PVC, so we have to convert it into rbdVol to access the child rbd image. It's done at a lot of places currently too.

Switched to using `conn.Copy()` so we should be good.
```
// Copy adds an extra reference count to the used ConnEntry and returns the
// *rados.Conn if it was found.
```
(Will be pushed in next pr update)

nixpanic replied:
> It's done at a lot of places currently too.

Yes, but that does not make it good practice :vomiting_face: 

> Switched to using conn.Copy() so we should be good.

Better, still ugly. But I think that is as good as we can do for now.

nixpanic commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,344 @@
+	// internalMaxResults is the maximum number of results to return in a single
+	// response. This is used to limit the number of changed blocks that needs
+	// to be stored in memory before sending a response to the stream.
+	internalMaxResults = 20
```
Isn't 20 very small?

Rakshith-R replied:
> Isn't 20 very small?

I'm fine with any value here on the lower
The CSI spec gives us freedom to send lower number than asked for.

We'll have a followup to make this configurable.

Rakshith-R replied:
Any suggestions ?

nixpanic replied:
I expect that the size of a single metadata for a changed block is relatively small. Start with something like 128, so that the number of repeated calls is reduced (will be less efficient?).

A configuration option is definitely good to do at a later point in time. Does the sidecar provide metrics about the number of changed blocks returned per diff? With that, a more educated guess for a default can be made.

Rakshith-R replied:
> I expect that the size of a single metadata for a changed block is relatively small. Start with something like 128, so that the number of repeated calls is reduced (will be less efficient?).

:+1: 

 
> A configuration option is definitely good to do at a later point in time. Does the sidecar provide metrics about the number of changed blocks returned per diff? With that, a more educated guess for a default can be made.

Currently only metrics regarding start & end time of requests is added 
@Nikhil-Ladha ^

nixpanic commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,344 @@
+	var fromSnapID uint64
+	image, err := rbdSnap.open()
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to open image %q: %s", rbdSnap, err)
```
`rbdSnap` functions should not return gRPC errors. gRPC errors should only be created in `ControllerServer` or other gRPC servers.

nixpanic reviewed (commented):
1st pass going through this. Needs cleanup and clear splitting of internal RBD things and ControllerServer Snapshot usage. Extend the `types.Snapshot` interface with required functions, so that ControllerServer does not need to use `rbdSnap` types directly.

nixpanic commented on PendingReleaseNotes.md:
```diff
@@ -4,6 +4,8 @@
 ## Features
 
+- rbd: add support for [CSI Snapshot Metadata Service RPCs](https://github.com/container-storage-interface/spec/blob/master/spec.md#snapshot-metadata-service-rpcs)
+
```
add something about it to the rbd documentation too?

nixpanic commented on internal/rbd/errors/errors.go:
```diff
@@ -68,3 +68,8 @@ var (
 )
+
+type ErrorCode interface {
+	// Code returns the error code.
```
Comment `Code` -> `ErrorCode`.

Can you also add a comment for the interface itself, with a small explanation how (which kind of errors) it can be used?

nixpanic commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,214 @@
+// StreamMetadata streams the block metadata for a snapshot.
+// If baseRBDSnap is provided, it calculates the delta between the
+// current snapshot and the base snapshot.
+func (rbdSnap *rbdSnapshot) StreamMetadata(
```
I have the feeling this function does too much. The `rbdSnapshot` struct should probably not send anything, but instead return `[]*csi.BlockMetadata` to the caller which can then send it out? Would that not make it a cleaner implementation?

Rakshith-R replied:
> I have the feeling this function does too much. 


Broke down this function to include multiple helpers

>The `rbdSnapshot` struct should probably not send anything, but instead return `[]*csi.BlockMetadata` to the caller which can then send it out? Would that not make it a cleaner implementation?

It's a streaming API, so we would be sending more than once.
Used this as inspiration to instead send down a func that only takes `[]*csi.BlockMetadata` as input and send stream response internally.

nixpanic replied:
Looks easier to understand already, thanks.

I understand that passing `sendResponse` makes things easier. It ties CSI+rbd functionality very closely together, which isn't very nice. But `librbd.DiffIterateByIDConfig{}` needs the callback, so I'm also not sure how to split the CSI/rbd things more. This is fine for now, maybe we can come up with an alternative approach later.

nixpanic commented on internal/rbd/types/snapshot.go:
```diff
@@ -37,4 +38,19 @@ type Snapshot interface {
+	// If baseRBDSnap is provided, the delta between the
+	// current snapshot and the base snapshot is streamed.
+	StreamMetadata(ctx context.Context,
+		stream grpc.ServerStream,
```
`grpc.ServerStream` really isn't nice here, see my other comment.

nixpanic commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,214 @@
+	}
+
+	if baseRBDSnap != nil {
+		fromSnapID, err = baseRBDSnap.GetRBDSnapID(ctx)
```
Instead of exporting the `GetRBDSnapID()` function in the Snapshot interface, it can be an internal function too. In this case, typecasting `baseRBDSnap` to a `*rbdSnapshot` would be ok.

Consider a helper function for typecasting:
```go
func rbdSnapFromSnapshot(snap types.Snapshot) (*rbdSnapshot, error) {}
```

nixpanic commented on internal/rbd/types/snapshot.go:
```diff
@@ -37,4 +38,19 @@ type Snapshot interface {
+	GetVolSize() int64
+
+	// GetRBDSnapID retrieves the RBD snapshot ID for the snapshot.
+	GetRBDSnapID(ctx context.Context) (uint64, error)
```
This should be a non-backend-specific naming, drop RBD from it. Ideally remove it from the interface, see a previous comment.

Rakshith-R committed:
doc: add documentation for CBT feature

Signed-off-by: Rakshith R <rar@redhat.com>

Madhu-1 commented on docs/rbd/deploy.md:
```diff
@@ -524,3 +524,36 @@ module is enabled on the nodes running ceph-csi attachers.
+
+   * Streams metadata about block differences between two snapshots
+   * Optimal for incremental backups
+   * Returns only blocks that changed between snapshots
```
This doc is for someone who wants to deploy and test it out, can we add relevant steps for the same?

Rakshith-R replied:
> This doc is for someone who wants to deploy and test it out, can we add relevant steps for the same?

Since ceph-csi-operator is the one deploying the sidecar, I'll add a link here to that doc when its ready over there.

Madhu-1 commented on internal/rbd/identityserver.go:
```diff
@@ -72,6 +73,25 @@ func (is *IdentityServer) GetPluginCapabilities(
+		caps = append(caps, &snapDiffCaps)
+	}
+	if err != nil {
+		log.DebugLog(ctx, "error checking for snapshot diff by ID support: %s", err)
```
log as error/warning?

Rakshith-R replied:
> log as error/warning?

Added it as warning since cephcsi can still function without these features.

Madhu-1 commented on internal/rbd/sms_controllerserver.go:
```diff
@@ -0,0 +1,220 @@
+
+	snapshotID := req.GetSnapshotId()
+
+	ctx := context.WithValue(stream.Context(), log.ReqID, req.GetSnapshotId())
```
log.ReqID will not be set for this one unless its added to middle ware? and also set the SnapshotID in the middleware itself?

Rakshith-R replied:
> log.ReqID will not be set for this one unless its added to middle ware? and also set the SnapshotID in the middleware itself?

Yes, its not logged :disappointed: ,

we need a new middle ware for streamserver interceptor, current one does not work these stream rpcs.
I'll send a separate pr for that.

Madhu-1 commented on internal/rbd/sms_controllerserver.go:
```diff
@@ -0,0 +1,220 @@
+	baseSnapshotID := req.GetBaseSnapshotId()
+	targetSnapshotID := req.GetTargetSnapshotId()
+
+	ctx := context.WithValue(stream.Context(), log.ReqID, targetSnapshotID)
+	ctx = context.WithValue(ctx, "Base Snapshot ID", baseSnapshotID)
```
set this in middleware?

Rakshith-R replied:
same as above
https://github.com/ceph/ceph-csi/pull/5347#discussion_r2166324902

Will add a separate pr for stream interceptor.

Madhu-1 replied:
Please open a tracker of it, so that we dont miss it

Madhu-1 commented on internal/rbd/sms_controllerserver.go:
```diff
@@ -0,0 +1,220 @@
+	defer mgr.Destroy(ctx)
+
+	baseRBDSnap, err := mgr.GetSnapshotByID(ctx, baseSnapshotID)
+	if errors.Is(err, rbderrors.ErrImageNotFound) {
+		log.ErrorLog(ctx, "snapshot %q not found: %v", baseSnapshotID, err)
+
+		return status.Errorf(codes.NotFound, "snapshot %q not found: %v", baseSnapshotID, err)
+	}
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
```
```suggestion
if err != nil {
    if errors.Is(err, rbderrors.ErrImageNotFound) {
		log.ErrorLog(ctx, "snapshot %q not found: %v", baseSnapshotID, err)

		return status.Errorf(codes.NotFound, "snapshot %q not found: %v", baseSnapshotID, err)
	}
		return status.Error(codes.Internal, err.Error())
	}

it would be nice to have one block for the error handling, take care of it in other places as well.

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,195 @@
+	defer func() {
+		cErr := image.Close()
+		if cErr != nil {
+			log.WarningLog(ctx, "resource leak, failed to close image: %v", cErr)
```
log it as error?

nixpanic replied:
I think of errors as fatal, a warning for a resource leak (unlikely to happen) is more suitable.

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,195 @@
+	diffIterateErr := image.DiffIterateByID(diffIterateByIDConfig)
+	err = handleDiffIterateError(ctx, diffIterateErr)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to get diff: %s", err)
```
log it at the caller? we dont need to log this at multiple places in this function?

Madhu-1 commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,195 @@
+// handleDiffIterateError checks the error returned by DiffIterateByID.
+// It handles specific error codes and returns a formatted error message
+// for unrecognized error codes.
+func handleDiffIterateError(ctx context.Context, err error) error {
```
let the caller log it and we dont need to pass the ctx?

Madhu-1 commented on internal/rbd/snapshot.go:
```diff
@@ -400,3 +400,48 @@ func (rbdSnap *rbdSnapshot) SetVolumeGroup(ctx context.Context, cr *util.Credent
+
+// rbdSnapFromSnapshot converts a types.Snapshot to a *rbdSnapshot.
+// It returns an error if the snapshot is nil or not of type *rbdSnapshot.
+func rbdSnapFromSnapshot(snap types.Snapshot) (*rbdSnapshot, error) {
```
can we add a function to Manager interface to return `rbdSnapshot` we dont need to have implementation here we can directly pass `*rbdSnapshot` to the `StreamMetadata` method

Rakshith-R replied:
> can we add a function to Manager interface to return `rbdSnapshot` we dont need to have implementation here we can directly pass `*rbdSnapshot` to the `StreamMetadata` method


>*rbdSnapshot to the StreamMetadata method
this was the implementation in one of my earlier iteration

The current approach was suggested by @nixpanic in 

https://github.com/ceph/ceph-csi/pull/5347#discussion_r2128818985

nixpanic replied:
> can we add a function to Manager interface to return rbdSnapshot

No, the idea of the manager is to separate the CSI components from the RBD internals. Clean APIs for all the things make it a much more maintainable project. Eventually...

Madhu-1 reviewed (commented):
some nits, overall LGTM

Nikhil-Ladha commented on docs/rbd/deploy.md:
```diff
@@ -527,7 +527,7 @@ If custom image is built for the rbd-plugin instance, make sure that it contains
 
 ## Changed Block Tracking (CBT)
 
-> **Warning**: Requires Ceph version that supports RBD snap diff by ID feature. 
```
These changes should be part of the previous `doc: add documentation for CBT feature` commit.

Nikhil-Ladha commented on internal/rbd/sms_controllerserver.go:
```diff
@@ -0,0 +1,220 @@
+	// defaultMaxResults is the internal limit for max results in metadata streaming
+	// This is set to a reasonable value to prevent excessive memory usage
+	// and ensure efficient processing of metadata requests.
+	// TODO: Consider making this configurable via a cmdline flag.
```
Please have an issue created for this, so it's easier to track.

Rakshith-R replied:
https://github.com/ceph/ceph-csi/issues/5401
created

Rakshith-R commented:
updated the code with
- modification as per reviews
- added logic to handle luksHeader if it exits by adding it to initial offset and subtracting it from offset that will be returned to client.

PTAL

Rakshith-R commented:
https://github.com/ceph/ceph-csi/actions/runs/15966099828/job/45026731182?pr=5347
```
//nolint:unused // Unused code will be used in future.
```
forgot to remove this nolint comment :face_exhaling: 
removed it and rebased on top of devel

PTAL

Nikhil-Ladha reviewed (commented):
lgtm

Nikhil-Ladha reviewed (commented):
LGTM

Madhu-1 commented:
@nixpanic Any final comments before merging as you have reviewed it

nixpanic commented on internal/rbd/identityserver.go:
```diff
@@ -72,6 +73,25 @@ func (is *IdentityServer) GetPluginCapabilities(
+
+	// RBDSnapDiffByID is used within the SnapshotMetadata service
+	rbdSnapDiffByIDSupported, err := features.SupportsRBDSnapDiffByID()
+	if err == nil && rbdSnapDiffByIDSupported {
```
you can write this a little simpler:

```go
	if err != nil {
		log.WarningLog(ctx, "error checking for snapshot diff by ID support: %s", err)
	} else if rbdSnapDiffByIDSupported {
		snapDiffCaps := csi.PluginCapability{
			Type: &csi.PluginCapability_Service_{
				Service: &csi.PluginCapability_Service{
					Type: csi.PluginCapability_Service_SNAPSHOT_METADATA_SERVICE,
				},
			},
		}
		caps = append(caps, &snapDiffCaps)
	}
```

nixpanic commented on internal/rbd/sms_controllerserver.go:
```diff
@@ -0,0 +1,226 @@
+	return requestedMaxResults
+}
+
+func validateMetadataAllocatedReq(req *csi.GetMetadataAllocatedRequest) error {
```
Consider including `SupportsRBDSnapDiffByID()` in the validate functions. VolumeGroupSnapshot calls still arrived at the controller, even when the support for it was not returned in the capabilities.

Or (cleaner?) check at the start of each procedure, like in `GetMetadataAllocated()`.

nixpanic commented on internal/rbd/sms_controllerserver.go:
```diff
@@ -0,0 +1,226 @@
+
+	startingOffset := req.GetStartingOffset()
+	if startingOffset >= volSize {
+		return status.Error(codes.OutOfRange, "starting offset is out of range")
```
consider including the requested offset/size

nixpanic commented on internal/rbd/sms_controllerserver.go:
```diff
@@ -0,0 +1,226 @@
+
+	sendResponse := func(changedBlocks []*csi.BlockMetadata) error {
+		return stream.Send(&csi.GetMetadataAllocatedResponse{
+			BlockMetadataType:   csi.BlockMetadataType_VARIABLE_LENGTH,
```
is this variable? I assumed RBD uses fixed sized block?

Rakshith-R replied:
> is this variable? I assumed RBD uses fixed sized block?

https://github.com/container-storage-interface/spec/blob/master/spec.md#metadata-format
```
The FIXED_LENGTH style requires a fixed value for the size_bytes field in all tuples in a given sequence.
The VARIABLE_LENGTH style does not constrain the value of the size_bytes field in a sequence.
```

```
/ # /snapshot-metadata-lister -max-results 10 -starting-offset 0 -namespace default -kubeconfig=“” -snapshot snapshot-1
Record#   VolCapBytes  BlockMetadataType   ByteOffset     SizeBytes
------- -------------- ----------------- -------------- --------------
      1      536870912   VARIABLE_LENGTH              0        4194304
/ # /snapshot-metadata-lister -max-results 10 -starting-offset 123213 -namespace default -kubeconfig=“” -snapshot snapshot-1
Record#   VolCapBytes  BlockMetadataType   ByteOffset     SizeBytes
------- -------------- ----------------- -------------- --------------
      1      536870912   VARIABLE_LENGTH         123213        4071091
```

ceph has fixed sized blocks for each object internally but sizeBytes is not constant in all cases. :thinking:

nixpanic commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,198 @@
+		sendResponse,
+		&changedBlocks,
+		maxResults,
+		luksHeaderPadding,
```
should this not be `startingOffset`?

Rakshith-R replied:
> should this not be `startingOffset`?

No, offset is passed below as part of config( 3 lines below).

luksHeaderPadding in order to adjust the offset sent to the user for encrypted volumes.

nixpanic commented on internal/rbd/snap_diff.go:
```diff
@@ -0,0 +1,198 @@
+	return func(offset, sizeBytes uint64, _ int, _ interface{}) int {
+		select {
+		case <-ctx.Done():
+			return int(codes.Canceled)
```
This callback should probably return go-ceph/rbd error values? Internally in go-ceph `getError(..)` is called, and that may convert some non-0 value as a real error. gRPC error codes are different from C errno values.

This returns a `librbd.DiffIterateCallback` function, so it needs to follow the librbd specs.

Rakshith-R replied:
> This callback should probably return go-ceph/rbd error values? Internally in go-ceph `getError(..)` is called, and that may convert some non-0 value as a real error. gRPC error codes are different from C errno values.
> 
> This returns a `librbd.DiffIterateCallback` function, so it needs to follow the librbd specs.

Actually the callback function is free to decide the error code, which is again interpreted by the user who has written the callback function.

librbd just calls this function and passed the error code as is to the caller if non-zero.

Rakshith-R replied:
Example in go-ceph tests https://github.com/ceph/go-ceph/blob/4209816079604c4f56adbf91db7ec466aca07bb4/rbd/diff_iterate_test.go#L321-L340

nixpanic replied:
ah, ok. And `codes.OK` is equal to 0?

nixpanic commented on internal/rbd/snapshot.go:
```diff
@@ -400,3 +400,48 @@ func (rbdSnap *rbdSnapshot) SetVolumeGroup(ctx context.Context, cr *util.Credent
 	return nil
 }
+
+// GetVolSize returns the size of the snapshot in bytes.
```
Just call this `GetSize()`, no need for the weird `Vol` part.

nixpanic commented on internal/rbd/types/snapshot.go:
```diff
@@ -25,6 +25,9 @@ import (
 )
 
+// SMSSendResponse is a function type for sending responses in the SMS controller server.
+type SMSSendResponse func([]*csi.BlockMetadata) error
```
maybe call this `MetadataCallback`, there is no direct relation to the SMS service (it is just the only thing using it, now)

nixpanic commented on internal/rbd/types/snapshot.go:
```diff
@@ -37,4 +40,16 @@ type Snapshot interface {
+	// StreamMetadata streams the block metadata for a snapshot.
+	// If baseRBDSnap is provided, the delta between the
+	// current snapshot and the base snapshot is streamed.
+	StreamMetadata(ctx context.Context,
```
like `SMSSendResponse`, maybe call this function `ProcessMetadata` or something more generic?

nixpanic commented on internal/rbd/snap_diff_test.go:
```diff
@@ -0,0 +1,108 @@
+	return e.code
+}
+
+func newMockError(code int, msg string) *mockError {
```
what about

```go
if code == 0 {
    return nil
}
```

That way, you can add a test case for `codes.OK` as well.

Rakshith-R replied:
> what about
> 
> ```go
> if code == 0 {
>     return nil
> }
> ```
> 
> That way, you can add a test case for `codes.OK` as well.

Added a codes.ok case inside handleDiffIterateError() function and corresponding test case is added too.

nixpanic commented on internal/rbd/types/snapshot.go:
```diff
@@ -37,4 +40,16 @@ type Snapshot interface {
+	// ProcessMetadata processes the block metadata for a snapshot.
+	// If baseRBDSnap is provided, the delta between the
+	// current snapshot and the base snapshot is processed.
+	ProcessMetadata(ctx context.Context,
```
a common coding practice is to have the callback as last parameter for functions (with optional callback arguments after that)

Rakshith-R replied:
:+1: ,callback function are at the end now.

Madhu-1 commented:
@Nikhil-Ladha PTAL as you have requested changes

Madhu-1 commented:
@mergifyio queue

Madhu-1 commented:
@Mergifyio rebase

Madhu-1 commented:
@mergifyio queue

Rakshith-R commented:
@mergifyio requeue
